rust/kernel/file_operations.rs

```diff
@@ -79,13 +79,26 @@ pub enum SeekFrom {
     Current(i64),
 }
 
+/// Called by the VFS when an inode should be opened.
+///
+/// Calls `T::open` on the returned value of `A::convert`.
+///
+/// # Safety
+///
+/// The returned value of `A::convert` must be a valid non-null pointer and
+/// `T:open` must return a valid non-null pointer on an `Ok` result.
 unsafe extern "C" fn open_callback<A: FileOpenAdapter, T: FileOpener<A::Arg>>(
     inode: *mut bindings::inode,
     file: *mut bindings::file,
 ) -> c_types::c_int {
     from_kernel_result! {
+        // SAFETY: `A::convert` must return a valid non-null pointer
         let arg = unsafe { A::convert(inode, file) };
+        // SAFETY: `arg` was returned by `A::convert` and must be a valid
+        // non-null pointer
         let ptr = T::open(unsafe { &*arg })?.into_pointer();
```

Comment:
Only the `unsafe { &*arg }` block needs the `// SAFETY` comment. To arrive at a correct comment here, ask yourself how you can prove that it is **always** safe to convert the raw pointer `arg` to a shared reference, which lives long enough to satisfy the requirements of `FileOpener::open()`.

(note that I do not know the answer to that question myself, this is intended to show what questions to ask and answer)

Reply:
Got it. I wrote it this way because I'm assuming the pointer returned by convert is going to point to something in the inode or file which will definitely live beyond open callback and will live beyond `T::open`. I'll add that to the comment.